crash on user quota reading

Terraform crash at user read step when it reads quotas informations.
One of the way to reproduce:
- create a user via terraform
- create a VM owned by the new user
- make a plan

Log:
```
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

opennebula_user.user: Refreshing state... [id=2]

Error: rpc error: code = Unavailable desc = transport is closing


panic: set item just set doesn't exist
2021-06-10T21:47:54.507+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0: 
2021-06-10T21:47:54.507+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0: goroutine 52 [running]:
2021-06-10T21:47:54.507+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0: github.com/hashicorp/terraform/helper/schema.(*MapFieldWriter).setSet(0xc00051c5a0, 0xc0004c0e40, 0x3, 0x3, 0xf0a820, 0xc00051c740, 0xc0000caf00, 0xc0001f7060, 0x410a24)
2021-06-10T21:47:54.507+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0:       /home/user/go/pkg/mod/github.com/hashicorp/terraform@v0.12.12/helper/schema/field_writer_map.go:328 +0xbfb
2021-06-10T21:47:54.507+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0: github.com/hashicorp/terraform/helper/schema.(*MapFieldWriter).set(0xc00051c5a0, 0xc0004c0e40, 0x3, 0x3, 0xf0a820, 0xc00051c3c0, 0x2, 0x2)
2021-06-10T21:47:54.507+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0:       /home/user/go/pkg/mod/github.com/hashicorp/terraform@v0.12.12/helper/schema/field_writer_map.go:107 +0x24a
2021-06-10T21:47:54.507+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0: github.com/hashicorp/terraform/helper/schema.(*MapFieldWriter).setObject(0xc00051c5a0, 0xc00051c640, 0x2, 0x2, 0xf8d020, 0xc0004c0b10, 0xc00012dc00, 0x20, 0xf1ee60)
2021-06-10T21:47:54.507+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0:       /home/user/go/pkg/mod/github.com/hashicorp/terraform@v0.12.12/helper/schema/field_writer_map.go:225 +0x227
2021-06-10T21:47:54.507+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0: github.com/hashicorp/terraform/helper/schema.(*MapFieldWriter).set(0xc00051c5a0, 0xc00051c640, 0x2, 0x2, 0xf8d020, 0xc0004c0b10, 0x1, 0x1)
2021-06-10T21:47:54.507+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0:       /home/user/go/pkg/mod/github.com/hashicorp/terraform@v0.12.12/helper/schema/field_writer_map.go:109 +0x1d8
2021-06-10T21:47:54.507+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0: github.com/hashicorp/terraform/helper/schema.(*MapFieldWriter).setList.func1(...)
2021-06-10T21:47:54.507+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0:       /home/user/go/pkg/mod/github.com/hashicorp/terraform@v0.12.12/helper/schema/field_writer_map.go:123
2021-06-10T21:47:54.507+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0: github.com/hashicorp/terraform/helper/schema.(*MapFieldWriter).setList(0xc00051c5a0, 0xc0006de3b0, 0x1, 0x1, 0xf0a2e0, 0xc00051c560, 0xc00012d600, 0xc0001f7388, 0x40f30d)
2021-06-10T21:47:54.507+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0:       /home/user/go/pkg/mod/github.com/hashicorp/terraform@v0.12.12/helper/schema/field_writer_map.go:143 +0x2b0
2021-06-10T21:47:54.507+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0: github.com/hashicorp/terraform/helper/schema.(*MapFieldWriter).set(0xc00051c5a0, 0xc0006de3b0, 0x1, 0x1, 0xf0a2e0, 0xc00051c560, 0x1, 0x91000000010a2740)
2021-06-10T21:47:54.507+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0:       /home/user/go/pkg/mod/github.com/hashicorp/terraform@v0.12.12/helper/schema/field_writer_map.go:103 +0xda
2021-06-10T21:47:54.507+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0: github.com/hashicorp/terraform/helper/schema.(*MapFieldWriter).WriteField(0xc00051c5a0, 0xc0006de3b0, 0x1, 0x1, 0xf0a2e0, 0xc00051c560, 0x0, 0x0)
2021-06-10T21:47:54.507+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0:       /home/user/go/pkg/mod/github.com/hashicorp/terraform@v0.12.12/helper/schema/field_writer_map.go:89 +0x3ef
2021-06-10T21:47:54.507+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0: github.com/hashicorp/terraform/helper/schema.(*MapFieldWriter).setSet(0xc00051c1a0, 0xc0006de3b0, 0x1, 0x1, 0xf0a2e0, 0xc00051c560, 0xc0000cb000, 0x0, 0x1e1871b0)
2021-06-10T21:47:54.507+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0:       /home/user/go/pkg/mod/github.com/hashicorp/terraform@v0.12.12/helper/schema/field_writer_map.go:307 +0x37e
2021-06-10T21:47:54.507+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0: github.com/hashicorp/terraform/helper/schema.(*MapFieldWriter).set(0xc00051c1a0, 0xc0006de3b0, 0x1, 0x1, 0xf0a2e0, 0xc00051c560, 0x1, 0x1)
2021-06-10T21:47:54.507+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0:       /home/user/go/pkg/mod/github.com/hashicorp/terraform@v0.12.12/helper/schema/field_writer_map.go:107 +0x24a
2021-06-10T21:47:54.507+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0: github.com/hashicorp/terraform/helper/schema.(*MapFieldWriter).WriteField(0xc00051c1a0, 0xc0006de3b0, 0x1, 0x1, 0xf0a2e0, 0xc00051c560, 0x0, 0x0)
2021-06-10T21:47:54.507+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0:       /home/user/go/pkg/mod/github.com/hashicorp/terraform@v0.12.12/helper/schema/field_writer_map.go:89 +0x3ef
2021-06-10T21:47:54.508+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0: github.com/hashicorp/terraform/helper/schema.(*ResourceData).Set(0xc000168af0, 0x10ec104, 0x6, 0xf0a2e0, 0xc00051c560, 0x0, 0x0)
2021-06-10T21:47:54.508+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0:       /home/user/go/pkg/mod/github.com/hashicorp/terraform@v0.12.12/helper/schema/resource_data.go:191 +0x1d8
2021-06-10T21:47:54.508+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0: github.com/OpenNebula/terraform-provider-opennebula/opennebula.resourceOpennebulaUserRead(0xc000168af0, 0x10df360, 0xc0006d23e0, 0xc000168af0, 0x0)
2021-06-10T21:47:54.508+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0:       /home/user/go/src/terraform-provider-opennebula/opennebula/resource_opennebula_user.go:187 +0x750
2021-06-10T21:47:54.508+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0: github.com/hashicorp/terraform/helper/schema.(*Resource).RefreshWithoutUpgrade(0xc0000c8480, 0xc0004bc820, 0x10df360, 0xc0006d23e0, 0xc0000b03b8, 0x0, 0x0)
2021-06-10T21:47:54.508+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0:       /home/user/go/pkg/mod/github.com/hashicorp/terraform@v0.12.12/helper/schema/resource.go:455 +0x119
2021-06-10T21:47:54.508+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0: github.com/hashicorp/terraform/helper/plugin.(*GRPCProviderServer).ReadResource(0xc0000b0048, 0x139ca20, 0xc000502bd0, 0xc0004bc640, 0xc0000b0048, 0xc000502bd0, 0xc000074a20)
2021-06-10T21:47:54.508+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0:       /home/user/go/pkg/mod/github.com/hashicorp/terraform@v0.12.12/helper/plugin/grpc_provider.go:525 +0x3d8
2021-06-10T21:47:54.508+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0: github.com/hashicorp/terraform/internal/tfplugin5._Provider_ReadResource_Handler(0x10aac60, 0xc0000b0048, 0x139ca20, 0xc000502bd0, 0xc0004bc5f0, 0x0, 0x139ca20, 0xc000502bd0, 0xc0005c0200, 0x74)
2021-06-10T21:47:54.508+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0:       /home/user/go/pkg/mod/github.com/hashicorp/terraform@v0.12.12/internal/tfplugin5/tfplugin5.pb.go:3153 +0x217
2021-06-10T21:47:54.508+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0: google.golang.org/grpc.(*Server).processUnaryRPC(0xc0006c2600, 0x13a7660, 0xc0006c3380, 0xc000166000, 0xc0001ca000, 0x1b1a3b0, 0x0, 0x0, 0x0)
2021-06-10T21:47:54.508+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0:       /home/user/go/pkg/mod/google.golang.org/grpc@v1.21.1/server.go:998 +0x46a
2021-06-10T21:47:54.508+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0: google.golang.org/grpc.(*Server).handleStream(0xc0006c2600, 0x13a7660, 0xc0006c3380, 0xc000166000, 0x0)
2021-06-10T21:47:54.508+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0:       /home/user/go/pkg/mod/google.golang.org/grpc@v1.21.1/server.go:1278 +0xd3d
2021-06-10T21:47:54.508+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0: google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0007001c0, 0xc0006c2600, 0x13a7660, 0xc0006c3380, 0xc000166000)
2021-06-10T21:47:54.508+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0:       /home/user/go/pkg/mod/google.golang.org/grpc@v1.21.1/server.go:717 +0xa1
2021-06-10T21:47:54.508+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0: created by google.golang.org/grpc.(*Server).serveStreams.func1
2021-06-10T21:47:54.508+0200 [DEBUG] plugin.terraform-provider-opennebula_v0.3.0:       /home/user/go/pkg/mod/google.golang.org/grpc@v1.21.1/server.go:715 +0xa1
2021/06/10 21:47:54 [ERROR] eval: *terraform.EvalRefresh, err: rpc error: code = Unavailable desc = transport is closing
2021-06-10T21:47:54.518+0200 [DEBUG] plugin: plugin process exited: path=.terraform/plugins/registry.terraform.io/opennebula/opennebula/0.3.0/linux_amd64/terraform-provider-opennebula_v0.3.0 pid=27935 error="exit status 2"
...
```

--- a/opennebula/quotas.go
+++ b/opennebula/quotas.go
@@ -235,7 +235,9 @@ func flattenQuotasMapFromStructs(d *schema.ResourceData, quotas *shared.QuotasLi
 		if qds.Size > 0 {
 			ds["size"] = qds.Size
 		}
-		datastoreQuotas = append(datastoreQuotas, ds)
+		if len(ds) > 0 {
+			datastoreQuotas = append(datastoreQuotas, ds)
+		}
 		q = q | F0
 	}
 	// Get network quotas
@@ -245,7 +247,9 @@ func flattenQuotasMapFromStructs(d *schema.ResourceData, quotas *shared.QuotasLi
 		if qn.Leases > 0 {
 			n["leases"] = qn.Leases
 		}
-		networkQuotas = append(networkQuotas, n)
+		if len(n) > 0 {
+			networkQuotas = append(networkQuotas, n)
+		}
 		q = q | F1
 	}
 	// Get VM quotas
@@ -272,7 +276,9 @@ func flattenQuotasMapFromStructs(d *schema.ResourceData, quotas *shared.QuotasLi
 		if quotas.VM.SystemDiskSize > 0 {
 			vm["system_disk_size"] = quotas.VM.SystemDiskSize
 		}
-		vmQuotas = append(vmQuotas, vm)
+		if len(vm) > 0 {
+			vmQuotas = append(vmQuotas, vm)
+		}
 		q = q | F2
 	}
 	// Get Image quotas
@@ -282,7 +288,9 @@ func flattenQuotasMapFromStructs(d *schema.ResourceData, quotas *shared.QuotasLi
 		if qimg.RVMs > 0 {
 			img["running_vms"] = qimg.RVMs
 		}
-		imageQuotas = append(imageQuotas, img)
+		if len(img) > 0 {
+			imageQuotas = append(imageQuotas, img)
+		}
 		q = q | F3
 	}
 

--- a/opennebula/quotas.go
+++ b/opennebula/quotas.go
@@ -10,10 +10,10 @@ import (
 )
 
 const (
-	F0 uint8 = 1
-	F1       = 2
-	F2       = 4
-	F3       = 8
+	DSFlag  uint8 = 1
+	NetFlag       = 2
+	VMFlag        = 4
+	ImgFlag       = 8
 )
 
 func quotasSchema() *schema.Schema {
@@ -238,7 +238,7 @@ func flattenQuotasMapFromStructs(d *schema.ResourceData, quotas *shared.QuotasLi
 		if len(ds) > 0 {
 			datastoreQuotas = append(datastoreQuotas, ds)
 		}
-		q = q | F0
+		q = q | DSFlag
 	}
 	// Get network quotas
 	for _, qn := range quotas.Network {
@@ -250,7 +250,7 @@ func flattenQuotasMapFromStructs(d *schema.ResourceData, quotas *shared.QuotasLi
 		if len(n) > 0 {
 			networkQuotas = append(networkQuotas, n)
 		}
-		q = q | F1
+		q = q | NetFlag
 	}
 	// Get VM quotas
 	if quotas.VM != nil {
@@ -279,7 +279,7 @@ func flattenQuotasMapFromStructs(d *schema.ResourceData, quotas *shared.QuotasLi
 		if len(vm) > 0 {
 			vmQuotas = append(vmQuotas, vm)
 		}
-		q = q | F2
+		q = q | VMFlag
 	}
 	// Get Image quotas
 	for _, qimg := range quotas.Image {
@@ -291,24 +291,24 @@ func flattenQuotasMapFromStructs(d *schema.ResourceData, quotas *shared.QuotasLi
 		if len(img) > 0 {
 			imageQuotas = append(imageQuotas, img)
 		}
-		q = q | F3
+		q = q | ImgFlag
 	}
 
 	quotasMap := make(map[string]interface{}, 0)
 	for q > 0 {
 		switch {
-		case q&F0 > 0:
+		case q&DSFlag > 0:
 			quotasMap["datastore_quotas"] = datastoreQuotas
-			q = q ^ F0
-		case q&F1 > 0:
+			q = q ^ DSFlag
+		case q&NetFlag > 0:
 			quotasMap["network_quotas"] = networkQuotas
-			q = q ^ F1
-		case q&F2 > 0:
+			q = q ^ NetFlag
+		case q&VMFlag > 0:
 			quotasMap["vm_quotas"] = vmQuotas
-			q = q ^ F2
-		case q&F3 > 0:
+			q = q ^ VMFlag
+		case q&ImgFlag > 0:
 			quotasMap["image_quotas"] = imageQuotas
-			q = q ^ F3
+			q = q ^ ImgFlag
 		}
 	}
 

